Calendario ubicación

JLeonN opened the pull request:
Después, díganme qué opinan
Debería el calendario estar ahí o en un lugar exclusivo para él?

📅 También, puede quedarse donde lo dejé, pero si le das clic, debería llevarte a una pestaña diferente donde se pueda ver en grande?

No sé, son ideas Después, cuéntenme.

JLeonN committed:
Calendario ubicación

Después, díganme qué opinan
Debería el calendario estar ahí o en un lugar exclusivo para él?

📅 También, puede quedarse donde lo dejé, pero si le das clic, debería llevarte a una pestaña diferente donde se pueda ver en grande?

No sé, son ideas Después, cuéntenme.

francosang commented:
Yo haria una nueva pagina y agregaría un link en el menú.
Así queda el calendario en su página.

Para que no quede todo tan cargado, porque va a tener dos botones para navegar de un mes a otro, y posiblemente se pueda hacer click en cada persona para ir a su perfil.

Para que todo tenga su lugar.

JLeonN committed:
Calendario ubicación - 2

De momento, va a estar acá. Luego veremos dónde dejarlo, o quizás agregamos un botón por ahí para ir directo al calendario. 🗓️

JLeonN committed:
Librería de calendario instalada

JLeonN committed:
Estilos del calendario listos

Incorporados estilos CSS al calendario y añadidas 3 pruebas de eventos.
Todo listo para la acción  📆

JLeonN committed:
CSS color de evento

francosang commented:
@JLeonN bien de bien boooo!

Creo que te confundiste e instalaste las dependencias en `/package.json` en lugar de en `/frontend/package.json`.
Acordate de que tenes que estar en el directorio `/frontend` antes de hacer el `npm install`.

Sería asi:
```shell
cd frontend
npm install dayjs
npm install react-big-calendar
```

francosang commented:
Queda re bien 🎉 👏 !!

![image](https://github.com/The-Cowboys/the-cowboy-day/assets/25726643/074f6dfe-1fa6-4360-968b-530c4cb10e0c)

francosang commented on frontend/src/Componentes/Calendario.jsx:
```diff
@@ -0,0 +1,48 @@
+    next: "Mes siguiente",
+  };
+
+  const events = [
+    {
+      start: dayjs("2024-05-09").toDate(),
+      end: dayjs("2024-05-11").toDate(),
+      title: "Prueba 1",
+    },
+    {
+      start: dayjs("2024-05-22").toDate(),
+      end: dayjs("2024-05-22").toDate(),
+      title: "Prueba 2",
+    },
+    {
+      start: dayjs("2024-06-22").toDate(),
+      end: dayjs("2024-06-22").toDate(),
+      title: "Prueba 3",
+    },
+  ];
```
Cuando ya no hagas mas pruebas, vas a tener que sacar los eventos de la API.
La nueva url es `/api/tontos/${año}/${mes}`, esto te devuelve una lista de tontos con el nombre y el dia que salio cada uno.


Para `start` y `end` vas a tener que usar el `dia` de cada tonto.
Y para el `title` vas a tener que usar el `nombre`.

francosang commented on frontend/src/Componentes/Calendario.jsx:
```diff
@@ -0,0 +1,48 @@
+  return (
+    <>
+      <div className="calendarioTamaño">
+        <Calendar
```
Cada vez que el mes cambia hay que hacer una nueva request a la API para obtener los tontos de ese mes.

Esta libería es un poco rara y no tiene una forma de obtener el mes selecionado, asi que no sé como vas a hacer.
Tampoco hay una forma de deshabilitar los botones que trae por defecto de Mes siguiente o Mes anterior. 

Lo unico que se me ocurre es tener un estado para el mes, que su valor inicial sea el mes actual, luego habría que ver si es posible escuchar los eventos de los botones para cambiar el valor del mes.
Pero no sé si la libreria lo dejara.

JLeonN replied:
Cuando terminé el calendario, me surgieron estas preguntas.
Por qué tuve problemas similares con el CSS, y luego me quedé pensando, cómo diablos le iba a poner un evento al botón? 🤔

Hasta se me cruzó la idea de cambiar de librería.
Luego te voy a mandar unos mensajes para hacerte unas preguntas un poco mejor.

francosang committed:
PRUEBAS DE FRANCO

JLeonN committed:
Api fecha

francosang commented on frontend/src/Componentes/Calendario.jsx:
```diff
@@ -0,0 +1,64 @@
+  const [eventos, setEventos] = useState([]);
+
+  useEffect(() => {
+    // TODO: como carajo obtener el rango de fechas inicial en el calendario?
```
@JLeonN una vez sepamos como obtener el rago de fechas visibles en el calendario, hay que llamar a la funcion `obtenerTontosEntreFechas(inicio, fin)`.

francosang commented on frontend/src/Componentes/Calendario.jsx:
```diff
@@ -0,0 +1,64 @@
+    // TODO 1: llamar a API para obtener el historial de tontos entre las dos fechas
+    // TODO 2: al recibir la lista de tontos del dia, mapearla a un arreglo de eventos
+
+    // ESTO ES UN EJEMPLO, GENERA NOMBRES ALEATOREOS
+    // BORRAR
+    const nombres = ["Jorge", "German", "Pablo", "Marcelo", "Cristian"];
+    const eventos = Array.from({ length: 31 }, (_, i) => {
+      return {
+        start: dayjs(inicio).add(i, "day").toDate(),
+        end: dayjs(inicio).add(i, "day").toDate(),
+        title: nombres[Math.floor(Math.random() * nombres.length)],
+      };
+    });
```
@JLeonN Esto lo hice para generar nombres aleatoreos. Hay que reemplazar todo esto por una llamada a la API de calendario.

francosang commented on frontend/src/Componentes/Calendario.jsx:
```diff
@@ -0,0 +1,64 @@
+          messages={messages}
+          views={["month"]}
+          events={eventos}
+          onRangeChange={(rango) => manejarCambioDeMes(rango)}
```
@JLeonN este evento lo dispara el componente de calendario.
En español el nombre sería algo como "cuando cambia el rango".

Significa que cada vez que cambia el mes visible en el calendario, se llama la funcion `manejarCambioDeMes(rango)`

francosang commented on frontend/src/Componentes/Calendario.jsx:
```diff
@@ -0,0 +1,64 @@
+    console.log("Fecha fin", fin);
+
+    // TODO 1: llamar a API para obtener el historial de tontos entre las dos fechas
+    // TODO 2: al recibir la lista de tontos del dia, mapearla a un arreglo de eventos
```
@JLeonN 

La API de calendario de tontos devuelve un objeto json que tiene esta forma:

```json
[
    {
        "id": 1,
        "nombre": "Pablo",
        "total": 7,
        "correo": "yoomat.75@hotmail.com",
        "dia": "2024-05-01T22:00:00.000Z"
    },
    {
        "id": 3,
        "nombre": "Franco",
        "total": 0,
        "correo": "el_mallic@hotmail.com",
        "dia": "2024-05-02T22:00:00.000Z"
    }
]
```

Los eventos del componente calendario necesitan esta forma:
```json
     {
        "start": fecha,
        "end": fecha,
        "title": titulo,
      }
```

Eso significa que hay que mapear (o transformar) los objetos de la API a los objetos que necesita el calendario.

JLeonN committed:
Obtener Tontos Del Mes

JLeonN committed:
Calendario - 1

franco-stuart commented on frontend/src/Componentes/Calendario.jsx:
```diff
@@ -0,0 +1,72 @@
+    // const inicio = ??????
+    // const fin = ??????
+    // obtenerTontosEntreFechas(inicio, fin)
+    // Alternativa
+    // - Otener fecha actual
+    // - Otener inicio del mes
+    // - Obtener fin del mes
+    // obtenerTontosEntreFechas(ini, f)
```
@JLeonN probaste estos pasos?

franco-stuart commented on package.json:
@JLeonN parece que estas librerias se instalaron en al proyecto de la API, en lugar de instalarse en el proyecto del frontend.

Cuando puedas deshace los cambios en este archivo e instala las dependencias en `frontend/package.json`.

JLeonN replied:
Sí, pero una pregunta: ¿cómo las desinstalo? Nunca desinstalé una librería. 🤔

franco-stuart commented on package-lock.json:
@JLeonN parece que estas librerias se instalaron en al proyecto de la API, en lugar de instalarse en el proyecto del frontend.

Cuando puedas revertite los cambios en este archivo e instala las dependencias en `frontend/package.json`. Cuando lo hagas, asegurate que estos cambios esten en `frontend/package-lock.json`.

franco-stuart reviewed (commented):
Me acabo de dar cuenta que las librerias del calendario se instalaron en la API 🤣 .
Deberían estar instaladas en el frontend.

Fijate de mover los cambios de `/package.json` a `/frontend/package.json` y de `/package-lock.json` a `/frontend/package-lock.json`

JLeonN committed:
Librerías fueron cambiadas de ubicación

Las librerías fueron cambiadas de ubicación.
Ahora están en frontend/package-lock.json.  📁

JLeonN committed:
Librerías desinstaladas

Desinstalé las librerías de la parte donde no iban y en el commit anterior ya las instalé en el frontend.  📦

JLeonN committed:
Arreglo de calendario

Franco, estuve con Pablo probando muchas cosas y terminamos poniendo un 'if'. No sé si es la mejor manera de resolverlo porque es la primera vez que lo hacemos, pero ahora el calendario funciona cuando presionas el botón de 'Hoy'. Solo me queda hacer lo del useEffect.

JLeonN committed:
Reversión de cambios

Franco, parece que en algún momento hice un 'Control Z' y cometí un error.

francosang committed:
Merge branch 'main' into Calendario

francosang committed:
terminar calendario
